### PR TITLE
#22: Fixed uninitialized mitigations

### LINKED ISSUE #22: Mitigations uninitialized when creating object with `new`
## Situation

The mitigation list is uninitialized when creating the object with `new` instead of create.

## Acceptance Criteria

1. Mitigation list initialized with empty list in when using `new`.

### PR DESCRIPTION
**Pull request by redcatbear**
Streamlined interface.
Updated handbook.
Fixed unitialized mitigations in case of `new`.

Closes #22.

Will update tests to busted in #23.

### CODE REVIEW

File: doc/user_guide/user_guide.md [hunk 1]
```diff
@@ -21,48 +21,52 @@ While this might sound counter-intuitive at first, all attributes of the error o
...
+```
+
+The `parameters` must have a field `value`. Optionally you can explain what a parameter means using the field `description`.
 
 ### Error Object Builder
 
-Alternatively you can use a builder provided with `exaerror` to construct an error object.
+Alternatively you can use builder-style construction provided with `ExaError` to create an error object.
 
 Strictly speaking it is a combination between the builder pattern and regular setters since Lua 5.1 does not have immutable tables.
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:66]: Do we still support Lua 5.1?

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -78,27 +82,27 @@ You can use the same parameter replacement mechanism in mitigations that you saw
 ```lua
 local project_issue_url = "www.example.org/issues"
 -- ...
-local msg = exaerror:new({
-    message = "Unexpected error.",
-    mitigations = {"Please create an error report under {{url}}."},
-    parameters = {url = project_issue_url}
-})
+local msg = ExaError:new(
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:85]:
```suggestion
local errobj = ExaError:new(
```

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -78,27 +82,27 @@ You can use the same parameter replacement mechanism in mitigations that you saw
...
-If you want to add a description to your parameters, you need to invest just a little bit more effort:
+If you want to add a description to your parameters, you need to invest just a bit more effort:
 
 ```lua
-local msg = exaerror:new({
-    message = "Unexpected error.",
-    mitigations = {"Please create an error report under {{url}}."},
-    parameters = {url = {value = project_issue_url, description = "URL under which you can raise issue tickets"}}
-})
+local msg = ExaError:new(
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:95]:
```suggestion
local errobj = ExaError:new(
```

File: doc/user_guide/user_guide.md [hunk 4]
```diff
@@ -107,7 +111,7 @@ Since we are talking about internal errors, that the users have no real chance o
 Use the following convenience method of the error object builder to cover those situations.
 
 ```lua
-local msg = exaerror.create("F-PACK-45", "Validation of created archive failed. Checksums do not match.")
+local msg = ExaError.create("F-PACK-45", "Validation of created archive failed. Checksums do not match.")
```
Comment (kaklakariada) [inline doc/user_guide/user_guide.md:114]:
```suggestion
local errobj = ExaError.create("F-PACK-45", "Validation of created archive failed. Checksums do not match.")
```

File: src/ExaError.lua [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+end
+
+--- Get the error code.
+-- @return error code
+function ExaError:get_code()
+    return self._code
+end
+
+--- Get the error message.
+-- This method supports Lua's standard string interpolation used in `string.format`.
```
Comment (kaklakariada) [inline src/ExaError.lua:97]: Is that true? `string.format` uses the `%s` syntax, not `{arg}`.

File: src/ExaError.lua [hunk 6]
```diff
@@ -0,0 +1,156 @@
+--- This class provides a uniform way to define errors in a Lua application.
+-- @module ExaError
```
Comment (kaklakariada) [inline src/ExaError.lua:2]:
```suggestion
-- @classmod ExaError
```

File: src/MessageExpander.lua [hunk 7]
```diff
@@ -1,77 +1,57 @@
...
-    self.tokens_ = {}
-    self.last_parameter_ = {characters = {}, quote = true}
-    return object
-end
-
----
--- Create new instance of a message expander.
---
+--- Create a new instance of a message expander.
 -- @parameter message to be expanded
```
Comment (kaklakariada) [inline src/MessageExpander.lua:13]:
```suggestion
-- @param message to be expanded
```

File: src/MessageExpander.lua [hunk 8]
```diff
@@ -1,77 +1,57 @@
...
--- @parameter ... values used to replace the placeholders
---
-function M.create(message, ...)
-    return M:new({
-        message = message,
-        parameters = {...},
-    })
+-- @parameter parameters parameter definitions
```
Comment (kaklakariada) [inline src/MessageExpander.lua:14]:
```suggestion
-- @param parameters parameter definitions
```

File: test/test_exaerror.lua [hunk 9]
```diff
@@ -1,97 +1,93 @@
+package.path = "src/?.lua;" .. package.path
+
 local luaunit = require("luaunit")
-local exaerror = require("exaerror")
+local exaerror = require("ExaError")
```
Comment (kaklakariada) [inline test/test_exaerror.lua:4]:
```suggestion
local ExaError = require("ExaError")
```
-> also rename in the rest of the file

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: doc/user_guide/user_guide.md (hunk 1 above)
Comment (redcatbear) [reply doc/user_guide/user_guide.md:66]: For this base module? Yes. It is useful outside of the Exasol context too.

File: src/ExaError.lua (hunk 5 above)
Comment (redcatbear) [reply src/ExaError.lua:97]: Not anymore. This is a left-over from a previous implementation that was more Lua-ish and less compliant with our other error-modules.

Comment (redcatbear) [reply src/ExaError.lua:97]: Removed the sentence.

File: doc/changes/changes_2.0.0.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+Release 2.0.0 of `error-reporting-lua` fixes a bug that caused an error when you used a boolean value as message parameter.
+
+We also streamlined the interface. The `create` factory method is now gone and the `new` method was simplified. This unfortunately results in a breaking change, which is also the reason we increased the major number.
+
+In the course of reworking the `new` method we also fixed a bug cause by an uninitialized mitigation in case the object was created with `new` instead of `create`.
+
+To comply with our [style guide](https://github.com/exasol/lua-coding/blob/main/doc/lua_style_guide.md) we changed the class names to upper camel case: `ExaError` and `MessageExpander`. Another breaking change.
+
+Of course, we updated the [user guide](../user_guide/user_guide.md) accordingly.
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/changes/changes_2.0.0.md:15]: Good idea!

File: doc/changes/changes_2.0.0.md [hunk 11]
```diff
@@ -8,12 +8,20 @@ Release 2.0.0 of `error-reporting-lua` fixes a bug that caused an error when you
...
+In the course of reworking the `new` method we also fixed a bug caused by an uninitialized mitigation in case the object was created with `new` instead of `create`.
 
 To comply with our [style guide](https://github.com/exasol/lua-coding/blob/main/doc/lua_style_guide.md) we changed the class names to upper camel case: `ExaError` and `MessageExpander`. Another breaking change.
 
 Of course, we updated the [user guide](../user_guide/user_guide.md) accordingly.
 
+### How to Migrate Your Code
+
+1. Replace `exaerror` by `ExaError` as module and class name
+2. If you used the message expander directly (which is usually not necessary), replace `message_expander` by `Message_Expander`
```
Comment (kaklakariada) [inline doc/changes/changes_2.0.0.md:20]:
```suggestion
2. If you used the message expander directly (which is usually not necessary), replace `message_expander` by `MessageExpander`
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

**Review by kaklakariada — DISMISSED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**